runtime/lua/vim/lsp.lua

```diff
@@ -662,6 +663,30 @@ function lsp.client()
   error()
 end
 
+--- Starts a new client or re-uses an existing client and attaches the current
+--- buffer to it.
+---
+---@param config table Same configuration as documented in |lsp.start_client()|
+---@param reuse_client nil|fun(client: table): boolean Function used to decide
+--- if a client should be re-used. All running clients are tested. Starts a new
+--- client if nothing matches. If `nil` clients that share the same name and
+--- same `root_dir` are re-used
+function lsp.start_or_attach(config, reuse_client)
+  reuse_client = reuse_client
+    or function(client)
+      return client.config.root_dir == config.root_dir and client.name == config.name
```

Comment:
wouldn't it be better if this already uses the streamline approach of `workspace_folders`? since `root_dir` is more or less a utility wrapper around it, imo

https://github.com/neovim/nvim-lspconfig/blob/9e6bcf5a8915e8423d5cc7f82c5069c11272184d/lua/lspconfig/configs.lua#L195-L201

Reply:
It would make the user setup more complicated for the "I have a single workspaceFolder" use-case and having this default implementation doesn't prevent a multi workspaceFolder use-case in the future.

As outlined in the PR description a future project extension could provide a different `reuse_client` function or we could even change the default implementation in a _mostly_ BWC way

It would also be possible to extend this later, so that if a user provides an alternative `reuse_client` function the `workspaceFolders` automatically get extended if the new config has a different `root_dir` (or different `workspaceFolders`.